hoomd/output/csv_backend.py

```diff
@@ -0,0 +1,295 @@
+
+        self._logger = logger
+        self._cur_headers_with_width = dict()
+        self._fmt = _Formatter(pretty, max_precision)
+        self._comm = None
+
+    def _setattr_param(self, attr, value):
+        """Makes self._param_dict attributes read only."""
+        raise ValueError("Attribute {} is read-only.".format(attr))
+
+    def attach(self, simulation):
+        self._comm = simulation.device._comm
+
+    def detach(self):
+        self._comm = None
+
+    def _get_log_dict(self):
+        """Get a flattened dict for writing to output."""
+        return {key: value[0]
+                for key, value in dict_flatten(self._logger.log()).items()
+                if value[1] in {'string', 'scalar'}
+                }
+
+    def _update_headers(self, new_keys):
+        """Update headers and write the current headers to output.
+
+        This function could be made simpler and faster by moving some of the
+        transformation to act. Since we don't expect the headers to change often
+        however, this would likely slow the writer down. The design is to
+        off-load any potnentially unnecessary calculations to this function even
+        if that means more overall computation when headers change.
+        """
+        header_output_list = []
+        header_dict = {}
+        for namespace in new_keys:
+            header = self._determine_header(
+                namespace, self.header_sep, self.max_header_len)
+            column_size = max(len(header), self.min_column_width)
+            header_dict[namespace] = column_size
+            header_output_list.append((header, column_size))
+        self._cur_headers_with_width = header_dict
+        self.output.write(
+            self.delimiter.join((self._fmt.format_str(hdr, width)
+                                  for hdr, width in header_output_list))
+        )
+        self.output.write('\n')
+
+    @staticmethod
+    def _determine_header(namespace, sep, max_len):
+        if max_len is None:
+            return sep.join(namespace)
+        else:
+            index = -1
+            char_count = len(namespace[-1])
+            for name in reversed(namespace[:-1]):
+                char_count += len(name)
+                if char_count > max_len:
+                    break
+                index -= 1
+            return sep.join(namespace[index:])
+
+    def _write_row(self, data):
+        """Write a row of data to output."""
+        headers = self._cur_headers_with_width
+        self.output.write(self.delimiter.join(
+            (self._fmt(data[k], headers[k]) for k in headers))
+                           )
+        self.output.write('\n')
+
+    def act(self, timestep=None):
+        """Write row to designated output.
+
+        Will also write header when logged quantites are determined to have
+        changed.
+        """
+        if self._comm is not None and self._comm.rank == 0:
+            output_dict = self._get_log_dict()
```

Comment:
This won't work for loggable properties that make MPI collective calls. Those properties need to be requested on all ranks. I think you can just move the `_get_log_dict` outside the `if`.

Reply:
Good catch!